core/drivers/stm32mp15_huk.c

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2022, Foundries.io Limited
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <crypto/crypto.h>
+#include <drivers/stm32_bsec.h>
+#include <kernel/tee_common_otp.h>
+#include <mempool.h>
+#include <platform_config.h>
+#include <stm32_util.h>
+#include <string.h>
+#include <string_ext.h>
+
+static bool stm32mp15_huk_init;
+
+static TEE_Result stm32mp15_read_uid(uint32_t *uid)
+{
+	uint32_t *q = uid;
+
+	/*
+	 * Shadow memory for UID words might not be locked: to guarante that
+	 * the final values are read we must lock them.
+	 */
+	if (stm32_bsec_set_sw_lock(UID0_OTP) ||
+	    stm32_bsec_shadow_read_otp(q++, UID0_OTP))
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_set_sw_lock(UID1_OTP) ||
+	    stm32_bsec_shadow_read_otp(q++, UID1_OTP))
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_set_sw_lock(UID2_OTP) ||
+	    stm32_bsec_shadow_read_otp(q++, UID2_OTP))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32mp15_read_otp(uint32_t otp, uint32_t *key, bool *locked)
+{
+	bool tmp = true;
+
+	if (!stm32mp_is_closed_device()) {
+		/*
+		 * When the device is not closed, the shadow memory for these
+		 * words might not be locked: check and report them
+		 */
+		if (stm32_bsec_read_permanent_lock(otp, &tmp))
+			return TEE_ERROR_GENERIC;
+
+		if (tmp && stm32_bsec_read_sw_lock(otp, &tmp))
+			return TEE_ERROR_GENERIC;
+	}
+
+	if (stm32_bsec_shadow_read_otp(key, otp))
+		return TEE_ERROR_GENERIC;
+
+	*locked = *locked && tmp;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ *  AES-GCM: nonce must be unique per message and key.
+ *
+ *  This function always uses the same key - once its locked - with the same
+ *  unique message hence the nonce can be any constant.
+ */
+static TEE_Result aes_gcm_encrypt_uid(uint8_t *key, size_t key_len,
+				      uint8_t *out, size_t *out_len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	const uint8_t nonce[12] = { 0x55 };
+	uint32_t uid[4] = { 0 };
+	uint8_t tag[16] = { 0 };
+	size_t nonce_len = sizeof(nonce);
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	void *ctx = NULL;
+
+	ret = stm32mp15_read_uid(uid);
+	if (ret)
+		goto err;
+
+	ret = crypto_authenc_alloc_ctx(&ctx, TEE_ALG_AES_GCM);
+	if (ret)
+		goto err;
+
+	ret = crypto_authenc_init(ctx, TEE_MODE_ENCRYPT, key, key_len, nonce,
+				  nonce_len, TEE_AES_BLOCK_SIZE, 0, uid_len);
+	if (ret)
+		goto auth_err;
+
+	ret = crypto_authenc_enc_final(ctx, (uint8_t *)uid, sizeof(uid),
+				       out, out_len, tag, &tag_len);
+	if (ret)
+		goto auth_err;
+
+	crypto_authenc_final(ctx);
+auth_err:
+	crypto_authenc_free_ctx(ctx);
+err:
```

Comment:
sorry I missed that. this is not a error only path so relabel to `out_free_ctx` annd `out`.

Reply:
sorry I'll never get used to this...ok will change,.